### PR TITLE
DCMAW-8229: Extract token verification logic

### PR DESCRIPTION
**Pull request by jhumbert-dd**
​DCMAW-8229

### What changed
- Extracted token verification logic into TokenVerifier class
- Updated handler and TokenService dependencies to include TokenVerifier
- Updated MockJWTBuilder
  - Added new function for building a singed JWT to enable testing of TokenVerifier
  - Amend JWT header property to be `typ` instead of `type`
- Updates asyncCredential lambda test to use new encoded header value to reflect change to header property key name

### Why did it change
To create better separation of concerns within token signature verification and to simplify adding and updating tests.

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [x] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [ ] Demo to a BA, TA, and the team.
- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ import { mockClient } from "aws-sdk-client-mock";
 import {
```
Comment (jmooney-dd) [inline backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts:12]: Do we need a test block for when `tokenVerifier.verify` returns an error result?

File: backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts [hunk 2]
```diff
@@ -22,6 +23,7 @@ describe("Token Service", () => {
   beforeEach(() => {
     dependencies = {
       publicKeyGetter: () => new MockPubicKeyGetterGetPublicKeySuccess(),
+      tokenVerifier: () => new MockTokenVerifierVerifySuccess(),
```
Comment (jmooney-dd) [inline backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts:26]: Do we need a test block for when tokenVerifier.verify returns an error result?

File: backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    jwt: string,
+    key: KeyLike | Uint8Array,
+  ): Promise<Result<JWTVerifyResult>> {
+    let result: JWTVerifyResult;
+    try {
+      result = await jwtVerify(jwt, key);
+    } catch (error) {
+      return errorResult({
+        errorMessage: `Error verifying token signature: ${error}`,
+        errorCategory: "SERVER_ERROR",
```
Comment (ahickmann-gds) [inline backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts:15]: This needs to be a client error, I believe. If we can't verify the signature then something is very likely wrong with the token.

File: backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+export class TokenVerifier implements ITokenVerifier {
+  async verify(
+    jwt: string,
+    key: KeyLike | Uint8Array,
+  ): Promise<Result<JWTVerifyResult>> {
+    let result: JWTVerifyResult;
+    try {
+      result = await jwtVerify(jwt, key);
+    } catch (error) {
```
Comment (ahickmann-gds) [inline backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts:12]: As much as I don't like saying this, can we not have `(error)` and then just return `errorMessage: "Error verifying token signature"...?

File: backend-api/src/functions/testUtils/mockJwtBuilder.ts [hunk 5]
```diff
@@ -83,6 +84,20 @@ export class MockJWTBuilder {
     return this;
   };
 
+  static getPublicKey = async (publicKey: IMockPublicKey) => {
+    return await importJWK(publicKey);
+  };
+
+  getSignedEncodedJwt = async (privateKey: IMockPrivateKey) => {
```
Comment (ahickmann-gds) [inline backend-api/src/functions/testUtils/mockJwtBuilder.ts:91]: I like this :D

File: backend-api/src/functions/testUtils/mockJwtBuilder.ts [hunk 6]
```diff
@@ -99,7 +114,7 @@ export class MockJWTBuilder {
 interface IMockJwt {
   header: {
     alg: string;
-    type: string;
+    typ: string;
```
Comment (ahickmann-gds) [inline backend-api/src/functions/testUtils/mockJwtBuilder.ts:117]: Good spot!

Comment (jhumbert-dd) [reply backend-api/src/functions/testUtils/mockJwtBuilder.ts:117]: Thank you!

File: backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts (hunk 3 above)
Comment (jhumbert-dd) [reply backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts]: Yes! Makes sense. Thanks - this has been updated now

File: backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts (hunk 4 above)
Comment (jhumbert-dd) [reply backend-api/src/functions/asyncActiveSession/tokenService/tokenVerifier.ts]: Yes - removed

File: backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts (hunk 2 above)
Comment (jhumbert-dd) [reply backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts:21]: Thanks @jmooney-dd. I've added this test

File: backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts [hunk 7]
```diff
@@ -352,7 +355,7 @@ describe("Token Service", () => {
     });
 
     describe("Token signature verification", () => {
-      describe("Given kid from JWT header is not found in JWKS", () => {
+      describe("Given there is an error getting public key from JWKS (e.g. kid from JWT header is not found in JWKS)", () => {
```
Comment (jhumbert-dd) [inline backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts]: I've renamed this test as it quite similar to the "Given there is an error retrieving the public key" test above. Any suggestions for a better description here?

Comment (ahickmann-gds) [reply backend-api/src/functions/asyncActiveSession/tokenService/tests/tokenService.test.ts]: It's fine line this for now, and it makes sense. We might have to change the tests a bit once we all moving parts come together.

**Review by ahickmann-gds — APPROVED** [review]

**Pull request merged**